### PR TITLE
Replaced ray_reduce with ray_map_reduce

### PR DESCRIPTION
**Pull request by danwallach**
The old code did the map in batches, manually, then called out to our own reduce library, which used `ray.wait`. This was a mess.

The new code has a general-purpose solution, where you define `map`, `reduce`, and `zero` methods on a "context" class, and all the rest is handled under the hood. In particular, rather than batching, we extend the original reduction strategy, of using `ray.wait` to pay attention to the number of pending tasks, and when it gets below the desired number, more map tasks are launched.

Thus, no batching, and we're always keeping as many tasks in flight, up to the desired ceiling, as possible.

All the tests pass, and it runs fine on localhost. Still need to try this on a big cluster. Also need a second pair of eyes to make sure I didn't do something stupid.

### CODE REVIEW

File: src/arlo_e2e/ray_map_reduce.py [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+
+    def __init__(
+        self,
+        context: MapReduceContext[T, R],
+        use_progressbar: bool = False,
+        input_description: str = DEFAULT_INPUT_STR,
+        reduction_description: str = DEFAULT_REDUCTION_STR,
+        max_tasks: int = 5000,
+        map_shard_size: int = 4,
+        reduce_shard_size: int = 10,
```
Comment (umbernhard) [inline src/arlo_e2e/ray_map_reduce.py:313]: it might be fun to play with automatic tuning of these. I'm sure they're a research paper out there discussing it.

File: tests/test_ray_map_reduce.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+
+    def zero(self) -> ElGamalCiphertext:
+        nonce = rand_q()  # we'll get a different zero each time
+        return self.map((nonce, 0))
+
+    def __init__(self, keypair: ElGamalKeyPair):
+        self.keypair = keypair
+
+
+class TestRayMapReduce(unittest.TestCase):
```
Comment (umbernhard) [inline tests/test_ray_map_reduce.py:54]: we should probably have tests for the corner cases, like 0 workers, 0 reduce shards, etc. to make sure the bounds assertions work. Also, check for inputs with only one item, etc. Though maybe we don't need to with hypothesis?

**Review by umbernhard — APPROVED** [review]
Most of the comments are documentation/not relevant to functionality, so I'm approving.

File: Pipfile [hunk 3]
```diff
@@ -15,13 +15,13 @@ pydocstyle = "*"
...
-hypothesis = "==5.43.3"
+pandas = '==1.1.5'
+hypothesis = "==6.1.1"
 typish = '==1.7.0'
 jsons = '==1.1.2'
-tqdm = "==4.47.0"
-cryptography = "==3.2"
-modin = '==0.8.0'
+tqdm = "==4.56.2"
+cryptography = "==3.3.2"
```
[the remark this answers is not in the document]
Comment (danwallach) [reply Pipfile:23]: Yeah, this was just adopting the "oops, there's a security hole" patch. Probably doesn't impact us, anyway.

File: src/arlo_e2e/memo.py [hunk 4]
```diff
@@ -75,15 +75,15 @@ def make_memo_lambda(func: Callable[[], Optional[T]]) -> Memo[T]:
...
-    the same object for `i` rather than each individual value.
-    >>> memos = [make_memo_lambda(lambda: i) for i in range(0, 10)]
-    >>> [memo.contents for memo in memos]
-    [9, 9, 9, 9, 9, 9, 9, 9, 9, 9]
-
-    Here's how you might make copies, to ensure everything works how you would expect.
-    >>> memos = [(lambda i: make_memo_lambda(lambda: i))(i) for i in range(0, 10)]
-    >>> [memo.contents for memo in memos]
-    [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+    the same object for `i` rather than each individual value::
```
[the remark this answers is not in the document]
Comment (danwallach) [reply src/arlo_e2e/memo.py:78]: This is a docstring blockquote feature, that I only recently discovered.

File: src/arlo_e2e/ray_helpers.py [hunk 5]
```diff
@@ -26,6 +26,7 @@ def ray_init_localhost(
         ray.init(num_cpus=num_cpus if num_cpus > 0 else os.cpu_count())
         _ray_is_local = True
         ray_post_init(write_failure_probability)
+        ray_wait_for_workers()
```
[the remark this answers is not in the document]
Comment (danwallach) [reply src/arlo_e2e/ray_helpers.py:29]: I actually threw that in there because I thought it was going to fix a spurious bug I was seeing. I'm flipping it to the same order as the remote case.

File: src/arlo_e2e/ray_helpers.py [hunk 6]
```diff
@@ -63,7 +64,11 @@ def ray_wait_for_workers(min_workers: int = 1) -> None:  # pragma: no cover
         return
 
     while True:
-        nodes = ray.workers()
+        # ray.workers() is indeed defined in Ray 1.1 or later, but for whateer
```
[the remark this answers is not in the document]
Comment (danwallach) [reply src/arlo_e2e/ray_helpers.py:67]: fixed

File: src/arlo_e2e/ray_map_reduce.py [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+from dataclasses import dataclass
+from ray import ObjectRef
+from ray.actor import ActorHandle
+
+from arlo_e2e.ray_progress import ProgressBar
+from arlo_e2e.utils import shard_list_uniform
+
+T = TypeVar("T")
+"""
+The input type for the map methods.
```
[the remark this answers is not in the document]
Comment (danwallach) [reply src/arlo_e2e/ray_map_reduce.py:17]: Done

File: src/arlo_e2e/ray_map_reduce.py [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+"""
+
+R = TypeVar("R")
+"""
+The output type for the map methods and the input/output type for the reduce methods.
```
[the remark this answers is not in the document]
Comment (danwallach) [reply src/arlo_e2e/ray_map_reduce.py:22]: Done

File: src/arlo_e2e/ray_map_reduce.py [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+    the Ray library or about the ordering in which it's executed.
+
+    Your instance of `RayMapReducer` includes this context as well as variables that define
+    exactly how the computation will be sharded up and distributed, whether there will be a
+    progress bar, and so forth.
+    """
+
+    _context: MapReduceContext[T, R]
+    _use_progressbar: bool
+    _input_description: str
```
[the remark this answers is not in the document]
Comment (danwallach) [reply src/arlo_e2e/ray_map_reduce.py:176]: They're all defined down in the `__init__` method. Should I repeat those definitions here?

File: src/arlo_e2e/ray_map_reduce.py (hunk 1 above)
Comment (danwallach) [reply src/arlo_e2e/ray_map_reduce.py:313]: Yeah, I thought about setting a timer, calling `ray.workers()` every so often, and automatically adjusting this thing. At some point, though, this whole map-reduce structure is something that the Ray people need to support themselves, since they know way better than we do exactly how much excess of ready-but-not-running tasks you need to keep around to keep the pipelines full.

File: tests/test_ray_map_reduce.py (hunk 2 above)
Comment (danwallach) [reply tests/test_ray_map_reduce.py:54]: Are we going for 100% code coverage? Because if we are, then there's a lot of extra stuff we need to test. Right now, we're floating somewhere around 80% overall.

File: src/arlo_e2e/ray_map_reduce.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (umbernhard) [reply src/arlo_e2e/ray_map_reduce.py:172]: Nope, I should just read more carefully haha

File: tests/test_ray_map_reduce.py (hunk 2 above)
Comment (umbernhard) [reply tests/test_ray_map_reduce.py:54]: In general, yes. it might be good to at least get good coverage in PRs adding new features. @benadida may have more thoughts about this.

Comment (benadida) [reply tests/test_ray_map_reduce.py:54]: yes I would like to get us to 100% test coverage for all things backend-y at least.
